### PR TITLE
issue#211 Add Lotus.env

### PR DESCRIPTION
**Pull request by nguyenngoc2505**
- Implement `Lotus.env`
```ruby
   Lotus.env # => "development"

  Lotus.env?(:development) # => true 
  Lotus.env?(:production)  # => false
```

### CODE REVIEW

File: lib/lotus/env.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+  # @example env = Lotus.env
+  #   # => #<Lotus::Env:0x000000024e56b8 @string="development">
+  #   # env.to_s
+  #     # "development"
+  #
+  #   # env.development?
+  #     # true
+
+  class Env < Utils::String
```
Comment (joneslee85) [inline lib/lotus/env.rb]: naming is hard, what do you think about renaming it to `EnvironmentInqurier` instead? I find `Env` vs `Environment` can cause confusion,

Comment (nguyenngoc2505) [inline lib/lotus/env.rb]: It sound good. Thanks you!

File: lib/lotus/environment_inqurier.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  # @example env = Lotus.env
+  #   # => #<Lotus::Env:0x000000024e56b8 @string="development">
+  #   # env.to_s
+  #     # "development"
+  #
+  #   # env.development?
+  #     # true
+
+  class EnvironmentInqurier < Utils::String
```
Comment (joneslee85) [inline lib/lotus/environment_inqurier.rb]: sorry I made a typo, it's supposed to be Inquirer not Inqurier

File: lib/lotus.rb [hunk 3]
```diff
@@ -9,4 +11,9 @@
 #
 # @see http://lotusrb.org
 module Lotus
+  # @since x.x.x
+
+  def self.env
+    @env ||= EnvironmentInqurier.new(Environment.new.environment)
```
Comment (jodosha) [inline lib/lotus.rb]: @nguyenngoc2505 This isn't thread safe. Can you please remove it the `@env ||=` part? Thank you :smile_cat:

File: lib/lotus/environment_inqurier.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    def production?
+      @string == PRODUCTION_ENV
+    end
+
+    def test?
+      @string == TEST_ENV
+    end
+
+    def method_missing(m, *args, &blk)
```
Comment (jodosha) [inline lib/lotus/environment_inqurier.rb]: binding `*args` and `&blk` isn't necessary as we don't use them.

**Comment by jodosha** [comment]
@nguyenngoc2505 I think string inquirer isn't necessary at all. We can have something like this:

```ruby
Lotus.env?(:production)
```

```ruby
module Lotus
  self.env?(name)
     # NOTE: Lotus::Environment doesn't have #environment? yet
     Lotus::Environment.new.environment?(name)
  end
end
```

**Comment by nguyenngoc2505** [comment]
@jodosha Thanks for your comment, I've updated it.

File: lib/lotus/environment.rb [hunk 5]
```diff
@@ -180,6 +180,13 @@ def environment
       @environment ||= ENV[LOTUS_ENV] || ENV[RACK_ENV] || DEFAULT_ENV
     end
 
+    # @since x.x.x
+    #
+    # @see Lotus.env?(name)
+    def environment?(name)
```
Comment (joneslee85) [inline lib/lotus/environment.rb]: can you write a test?

File: test/env_test.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+describe 'Lotus' do
+
+  describe '.env' do
+    it 'returns env object' do
+      Lotus.env.must_be_kind_of String
+    end
+  end
+
+  describe '.env.test?' do
```
Comment (joneslee85) [inline test/env_test.rb]: I think this should be `env?`

File: test/env_test.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    before do
+      ENV['LOTUS_ENV']  = 'test'
+    end
+
+    it 'return true' do
+      Lotus.env?(:test).must_equal true
```
Comment (joneslee85) [inline test/env_test.rb]: can we also test for false scenario?

File: test/env_test.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
+require 'test_helper'
+
+describe 'Lotus' do
+
+  describe '.env' do
+    it 'returns env object' do
+      Lotus.env.must_be_kind_of String
```
Comment (joneslee85) [inline test/env_test.rb]: can we assert that it returns `test` here?

File: lib/lotus/environment.rb (hunk 5 above)
Comment (nguyenngoc2505) [inline lib/lotus/environment.rb]: I've written. Thanks!

File: test/environment_test.rb [hunk 9]
```diff
@@ -118,6 +118,32 @@
     end
   end
 
+  describe '#environment?' do
+    describe 'environment?(:test)' do
```
Comment (joneslee85) [inline test/environment_test.rb]: What do you think about removing this line and change the 2 context description line to `when environment is matched` and `when environment is not matched` instead?

File: test/env_test.rb [hunk 10]
```diff
@@ -0,0 +1,33 @@
+require 'test_helper'
+
+describe 'Lotus' do
+  before do
+    ENV['LOTUS_ENV'] = 'test'
+  end
+
+  describe '.env' do
+    it 'returns env object' do
+      Lotus.env.must_be_kind_of String
```
Comment (joneslee85) [inline test/env_test.rb]: IMO, we do not need this assertion, it does not add any value

File: test/env_test.rb [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    end
+
+    it 'return "test"' do
```
Comment (joneslee85) [inline test/env_test.rb]: s/return/returns

File: test/env_test.rb [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+      Lotus.env.must_equal 'test'
+    end
+  end
+
+  describe '.env?(:test)' do
```
Comment (joneslee85) [inline test/env_test.rb]: what do you think about changing it to `.env?` and adding a context line next to it `when environment is matched`?

File: test/env_test.rb [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    it 'return true' do
```
Comment (joneslee85) [inline test/env_test.rb]: s/return/returns

File: test/env_test.rb [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+      Lotus.env?(:test).must_equal true
+    end
+
+    describe 'when environment is not "test"' do
+      before do
+        ENV['LOTUS_ENV'] = 'development'
+      end
+
+      it 'return false' do
```
Comment (joneslee85) [inline test/env_test.rb]: s/return/returns

File: lib/lotus.rb [hunk 15]
```diff
@@ -9,4 +10,18 @@
 #
 # @see http://lotusrb.org
 module Lotus
+
+  # @since x.x.x
+  #
```
Comment (joneslee85) [inline lib/lotus.rb]: What do you think about add description for this method, something like `Return the current environment`

File: lib/lotus.rb [hunk 16]
```diff
@@ -9,4 +10,18 @@
...
+  # @example Lotus.env => "development"
+  def self.env
+    Environment.new.environment
+  end
+
+  # @since x.x.x
```
Comment (joneslee85) [inline lib/lotus.rb]: What do you think about adding description for this method? e.g.: `Check to see if specified environment matches the current environment`

File: lib/lotus.rb (hunk 15 above)
Comment (nguyenngoc2505) [inline lib/lotus.rb]: Added. Thanks!

Comment (joneslee85) [inline lib/lotus.rb]: Should be on top of the since

File: test/env_test.rb [hunk 17]
```diff
@@ -0,0 +1,31 @@
+require 'test_helper'
+
+describe 'Lotus' do
+  before do
+    ENV['LOTUS_ENV'] = 'test'
+  end
+
+  describe '.env' do
+    it 'returns "test"' do
```
Comment (joneslee85) [inline test/env_test.rb]: What do you think about changing this to "returns environment name"?

File: lib/lotus.rb (hunk 16 above)
Comment (joneslee85) [inline lib/lotus.rb]: Move the description to top of other lines

File: test/env_test.rb (hunk 17 above)
Comment (nguyenngoc2505) [inline test/env_test.rb]: For more detail I think it will be `returns curent environment name`

**Comment by jodosha** [comment]
@nguyenngoc2505 This looks great, thank you :clap: :sparkles: 

May I ask to extra bonus points?

  1. Please test `Lotus.env?` and `Lotus::Environment#environment?` with strings, not just symbols.
  2. What do you think if we let `Lotus.env?` and `Lotus::Environment#environment?` to accept multiple names? It would be handy to check multiple envs at once (see the example below).

### Multiple envs check

Imagine this scenario

```ruby
if Lotus.env?(:staging) || Lotus.env?(:production)
  # do something
end
```

We can provide a shorter version

```
if Lotus.env?(:staging, :production)
  # do something
end
```

Please note that single name should still work.

**Comment by joneslee85** [comment]
@jodosha great suggestion
@nguyenngoc2505 hint: please look into how to use the single splat parameter

**Comment by nguyenngoc2505** [comment]
@jodosha , @joneslee85 Thanks for your suggestion. And #1, Do we accept or reject the `strings` variable? I think we accept the `strings` variable is good.

**Comment by joneslee85** [comment]
@nguyenngoc2505 we would accept String

**Comment by nguyenngoc2505** [comment]
@jodosha, @joneslee85 I've fixed this PR with your suggestion.

**Comment by joneslee85** [comment]
:+1:

**Pull request merged**